containers/src/Kokkos_DualView.hpp

```diff
@@ -1142,23 +1157,26 @@ class DualView : public ViewTraits<DataType, Arg1Type, Arg2Type, Arg3Type> {
 namespace Kokkos {
 namespace Impl {
 
-template <class D, class A1, class A2, class A3, class... Args>
-struct DualViewSubview {
+template <class DV, class... Args>
+struct DualViewSubviewType {
+  static_assert(Kokkos::is_dual_view_v<DV>);
+
   using dst_traits = typename Kokkos::Impl::ViewMapping<
-      void, Kokkos::ViewTraits<D, A1, A2, A3>, Args...>::traits_type;
+      void, typename DV::traits, Args...>::traits_type;
 
   using type = Kokkos::DualView<
       typename dst_traits::data_type, typename dst_traits::array_layout,
-      typename dst_traits::device_type, typename dst_traits::memory_traits>;
+      typename dst_traits::device_type, typename dst_traits::hooks_policy,
```

Comment:
Comment about that change.

Reply:
I'm not sure if you mean commenting that in the code or in the PR...

I'll comment it here first.

The whole point of this PR is to allow a `Kokkos::DualView` to accept any `Properties...` that a `Kokkos::View` can accept. Before this PR, it was hardcoded that `DualView` only accepts 4 types:
1. `DataType`
2. `Arg1Type`
3. `Arg2Type`
4. `Arg3Type`

but there was no check enforcing the type of the last 3 (Arg1Type, Arg2Type, Arg3Type). So they could be anything that `Kokkos::ViewTraits` would accept. However, `DualViewSubview::type` would then "hardcode" that when taking a subview of a dual view, only data type, array layout, device type and memory traits would be set. If my understanding is correct, that could have lead to issues if for instance I had created my original dual view with some hooks policy and then took a subview of it because my subview would have the default hooks policy (please correct me if this isn't correct! :wink:).

However, I must admit that I'm not satisfied of my solution because I only added `hooks_policy` while `Kokkos::ViewTraits` may also accept `specialize` and others.

So my question for you @dalg24 is how I can ensure that all traits are passed to `DualViewSubviewType::type`? I guess ideally one should be able to do:
```c++
template <class DV, class... Args>
struct DualViewSubviewType {
  static_assert(Kokkos::is_dual_view_v<DV>);

  using dst_traits = typename Kokkos::Impl::ViewMapping<
      void, typename DV::traits, Args...>::traits_type;

  using type = Kokkos::DualView<dst_traits>;
};
```

where `dst_traits` is a `Kokkos::ViewTraits` that would then embed all the traits given by the user and this would ensure that all types are passed. But I agree this would not work so I wonder how one can ensure that `DualViewSubviewType::type` will always "forward" all types, even in the future when `ViewTraits` would accept some more types...

Reply:
I would have tried something like
```C++
namespace Impl {

template <class V>
struct V2DV;

template <class D, class... P>
struct V2DV<View<D, P...>> {
  using type = DualView<D, P...>;
};
} /* namespace Impl */

template <class DataType, class... Properties, class... Args>
auto subview(const DualView<DataType, Properties...>& src, Args... args) {
  // leverage Kokkos::View facilities to deduce the properties of the subview
  using deduce_subview_type =
      decltype(subview(std::declval<View<DataType, Properties...>>(), args...));
  // map it back to dual view
  return typename Impl::V2DV<deduce_subview_type>::type(src, args...);
}
```
to defer the deduction to Kokkos Core.
Ping @nmm0 who introduced the view hooks and @masterleinad who recently refactored `subview(View<...>, ...)`

I creeped for "subview" in the containers unit test and the test coverage looks pretty light.

Reply:
You can use perfect forwarding if you like.